analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -29,12 +31,112 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors
+                OperationKindEx.RecursivePattern => BoolConstraintFromRecursivePattern(valueConstraint, As(IRecursivePatternOperationWrapper.FromOperation)),
+                OperationKindEx.DeclarationPattern => BoolConstraintFromDeclarationPattern(valueConstraint, As(IDeclarationPatternOperationWrapper.FromOperation)),
+                OperationKindEx.TypePattern when
+                    As(ITypePatternOperationWrapper.FromOperation) is var type
+                    && type.InputType.DerivesOrImplements(type.NarrowedType) => BoolConstraint.From(valueConstraint == ObjectConstraint.NotNull),
+                OperationKindEx.NegatedPattern => BoolConstraintFromPattern(state, valueConstraint, As(INegatedPatternOperationWrapper.FromOperation).Pattern)?.Opposite,
+                OperationKindEx.DiscardPattern => BoolConstraint.True,
+                OperationKindEx.BinaryPattern => BoolConstraintFromBinaryPattern(state, valueConstraint, As(IBinaryPatternOperationWrapper.FromOperation)),
+                _ => null,
+            };
+
+            T As<T>(Func<IOperation, T> fromOperation) =>
+                 fromOperation(pattern.WrappedOperation);
+        }
+
+        private static BoolConstraint BoolConstraintFromDeclarationPattern(ObjectConstraint valueConstraint, IDeclarationPatternOperationWrapper declaration) =>
+            declaration switch
+            {
+                { MatchesNull: true } => BoolConstraint.True,
+                _ when valueConstraint == ObjectConstraint.Null => BoolConstraint.False,
+                _ when declaration.InputType.DerivesOrImplements(declaration.NarrowedType) => BoolConstraint.From(valueConstraint == ObjectConstraint.NotNull),
+                _ => null,
+            };
+
+        private static BoolConstraint BoolConstraintFromRecursivePattern(ObjectConstraint valueConstraint, IRecursivePatternOperationWrapper recursive) =>
+            recursive switch
+            {
+                _ when valueConstraint == ObjectConstraint.Null => BoolConstraint.False,
+                {
+                    PropertySubpatterns.Length: 0,
+                    DeconstructionSubpatterns.Length: 0,
+                } => BoolConstraint.True,
+                _ when SubPatternsAlwaysMatch(recursive) => BoolConstraint.True,
+                _ => null,
+            };
+
+        private static bool SubPatternsAlwaysMatch(IRecursivePatternOperationWrapper recursivePattern) =>
+            recursivePattern.PropertySubpatterns
+                .Select(x => IPropertySubpatternOperationWrapper.FromOperation(x).Pattern)
+                .Concat(recursivePattern.DeconstructionSubpatterns.Select(x => IPatternOperationWrapper.FromOperation(x)))
+                .All(x => x is { WrappedOperation.Kind: OperationKindEx.DiscardPattern }
+                    || (x.WrappedOperation.Kind == OperationKindEx.DeclarationPattern
+                        && IDeclarationPatternOperationWrapper.FromOperation(x.WrappedOperation).MatchesNull));
```

Comment:
Just formatting, to make the `a || (b && c)` a bit less confusing
```suggestion
                .All(x => x is { WrappedOperation.Kind: OperationKindEx.DiscardPattern }
                    || (x.WrappedOperation.Kind == OperationKindEx.DeclarationPattern && IDeclarationPatternOperationWrapper.FromOperation(x.WrappedOperation).MatchesNull));
```